Explanatory Text Update (#11)
cfrg/draft-irtf-cfrg-bbs-pseudonyms: draft-irtf-cfrg-bbs-per-verifier-linkability.md

In some applications, however, a Verifier needs to track the presentations made by the Prover over time, as to provide security monitoring, monetization services, configuration persistance etc.. In other applications a Prover may wish to assert a pseudonymous identity associated with a signature from an issuer. To promote privacy, the Prover should not be forced to reveal or be bound to a unique identifier that would remain constant across proof presentations to different Verifiers and which could be used to link a Provers interactions with different Verifiers.

The goal of this document is to provide a way for a Verifier to track the proof presentations that are intended for them, while at the same time not allowing the tracking of the Prover's activities with other Verifiers. This is done through the use of a cryptographic pseudonyms.

> A cryptographic pseudonym, or pseudonym for short, as defined by this document, is a value that will be computed from two parts. One part is the pseudonym secret value (nym_secret) which is known only to the prover and a context value (context_id) that must be known by both prover and verifier. The pseudonym value is computed in such a way that it is computationally infeasable to link to pseudonyms to the same pseudonym secret, i.e., holder for two different context values.

Reviewer: ```suggestion
A cryptographic pseudonym, or pseudonym for short, as defined by this document, is a value that will be computed from two parts. One part is the pseudonym secret value (`nym_secret`) which is known only to the prover and a context value (`context_id`) that must be known by both prover and verifier. The pseudonym value is computed in such a way that it is computationally infeasable to link to pseudonyms to the same pseudonym secret, i.e., holder for two different context values.
```

## Pseudonyms Bound to BBS Signatures

> The BBS signature scheme is based on a three party model of *signer* (aka issuer), *prover* (aka user or holder), and *verifier*.  A *prover* obtains a BBS signature from a *signer* over a list of BBS *messages* and presents a BBS proof (of signature) along with a selectively disclosed subset of the BBS *messages* to a verifier. Each BBS proof generated is unlinkable to other BBS proofs derived from the same signature and from the BBS signature itself. If the disclosed subset of BBS *messages* are not linkable then the presentations cannot be linked.

Reviewer: ```suggestion
The BBS signature scheme is based on a three party model of *signer* (aka issuer), *prover* (aka user or holder), and *verifier*.  A *prover* obtains a BBS signature from a *signer* over a list of *messages* and presents a BBS proof (of signature possession) along with a selectively disclosed subset of the BBS *messages* to a verifier. Each BBS proof generated is unlinkable to other BBS proofs derived from the same signature and from the BBS signature itself. If the disclosed subset of BBS *messages* are not linkable then the presentations cannot be linked.
```

BBS pseudonyms extend the BBS signature scheme to "bind" a "pseudonym secret" to a BBS signature retaining all the properties of the BBS signature scheme: (a) a short signature over multiple messages, (b) selective disclosure of a subset of messages from *prover* to *verifier*, (c) unlinkable proofs.

In addition BBS pseudonyms provide for:

> 1. A essentially unique identifier, a pseudonyms, bound to a proof of signature whose linkability is under the control of the *prover* in conjunction with a *verifier* via the selection of a "context". Such a pseudonym can be used when a *prover* revisits a *verifier* to allow a *verifier* to recognize the prover when they return or for the *prover* to assert their pseudononous identity when visiting a *verifier*

Reviewer: ```suggestion
1. A essentially unique identifier, a pseudonym, bound to a proof of signature whose linkability is under the control of the *prover* in conjunction with a *verifier* via the selection of a "context". Such a pseudonym can be used when a *prover* revisits a *verifier* to allow a *verifier* to recognize the prover when they return or for the *prover* to assert their pseudononous identity when visiting a *verifier*
```

2. Assurance of per *signer* uniqueness of the "pseudonym secret", i.e., the *signer* assures that the pseudonyms that will be guaranteed by the signature have not been used with any other signature issued by the signer (unless a signature is intentionally reissued).
3. The *signer* cannot track the *prover* presentations to *verifiers* based on pseudonym values.
4. Colluding *verifiers* sharing BBS proofs with pseudonyms cannot link proofs or pseudonyms across "contexts".

> To realize the above feature set we embed a two part pseudonym capability into the BBS signature scheme. The pseudonym's cryptographic value will be computed from a secret part, which we call the *nym_secret* and a part that is public or at least shared between the *prover* and one or more *verifiers*. The public part we call the *context_id*. The pseudonym is calculated from these two pieces using discrete exponentiation. This is similar to the computations in [@Lys00] and [@ABC14]. The pseudonym is presented to the *verifier* along with a ZKP that the *prover* knows the *nym_secret* and used it and the *context_id* to compute the pseudonym value. A similar proof mechanism was used in [@Lys00].

Reviewer: ```suggestion
To realize the above feature set we embed a two part pseudonym capability into the BBS signature scheme. The pseudonym's cryptographic value will be computed from a secret part, which we call the *`nym_secret`* and a part that is public or at least shared between the *prover* and one or more *verifiers*. The public part we call the *`context_id`*. The pseudonym is calculated from these two pieces using discrete exponentiation. This is similar to the computations in [@Lys00] and [@ABC14]. The pseudonym is presented to the *verifier* along with a ZKP that the *prover* knows the *`nym_secret`* and used it and the *`context_id`* to compute the pseudonym value. A similar proof mechanism was used in [@Lys00].
```

> To bind a pseudonym to a BBS signature we have the *signer* utilzed Blind BBS signatures and essentially sign over a commitment to the *nym_secret*. Hence only a prover that knows the *nym_secret* can generate a BBS proof from the signature (and also generate the pseudonym proof).

Reviewer: ```suggestion
To bind a pseudonym to a BBS signature we have the *signer* utilzed Blind BBS signatures and essentially sign over a commitment to the *`nym_secret`*. Hence only a prover that knows the *`nym_secret`* can generate a BBS proof from the signature (and also generate the pseudonym proof).
```

> As in [@Lys00] we are concerned with the possibility of a dishonest user and hence require that that the *nym_secret* = *prover_nym* + *signer_nym_entropy* be the sum of two parts where the *prover_nym* is a provers secret and only sent to the *signer* in a binding and hiding commitment. The *signer_nym_entropy* is "blindly added" in by the *signer* during the signing procedure and sent back to the *prover* along with the signature. Note the order of operations. The *prover* chooses their (random) *prover_nym* and commits to it. They then send the commitment along with a ZKP proof that the *prover_nym* makes this commitment. The *signer* verifies the commitment to the *prover_nym* then generates the *signer_nym_entropy* and "blindly adds" it to the *prover_nym* during the signature process. Note that this can be done since we sign over the commitment and we know the generator for the commitment.

Reviewer: ```suggestion
As in [@Lys00] we are concerned with the possibility of a dishonest user and hence require that the *`nym_secret`* = *`prover_nym`* + *`signer_nym_entropy`* be the sum of two parts where the *`prover_nym`* is a provers secret and only sent to the *signer* in a binding and hiding commitment. The *`signer_nym_entropy`* is "blindly added" in by the *signer* during the signing procedure and sent back to the *prover* along with the signature. Note the order of operations. The *prover* chooses their (random) *`prover_nym`* and commits to it. They then send the commitment along with a ZKP proof that the commitment is correctly calculated. The *signer* verifies the commitment to the *`prover_nym`* then generates the *`signer_nym_entropy`* and "blindly adds" it to the *`prover_nym`* during the signature process. Note that this can be done since we sign over the commitment and we know the generator for the commitment.
```
